src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -1620,6 +1624,41 @@ because xbuild doesn't support framework reference assemblies.
   </ItemGroup>
 </Target>
 
+<Target Name="_PrepareAndroidRemapNativeAssemblySources">
+  <PrepareAbiItems
+      BuildTargetAbis="@(_BuildTargetAbis)"
+      NativeSourcesDir="$(_NativeAssemblySourceDir)"
+      InstantRunEnabled="$(_InstantRunEnabled)"
+      Debug="$(AndroidIncludeDebugSymbols)"
+      Mode="jniremap">
+    <Output TaskParameter="AssemblySources" ItemName="_AndroidRemapAssemblySource" />
+  </PrepareAbiItems>
+</Target>
+
+<Target Name="_GenerateEmptyAndroidRemapNativeCode"
+        DependsOnTargets="_PrepareAndroidRemapNativeAssemblySources"
+        Condition=" '@(_AndroidRemapMembers->Count())' == '0' "
+        Inputs="$(MSBuildProjectFullPath)"
+        Outputs="@(_AndroidRemapAssemblySource)">
+  <GenerateJniRemappingNativeCode
+      OutputDirectory="$(_NativeAssemblySourceDir)"
+      GenerateEmptyCode="True"
+      SupportedAbis="@(_BuildTargetAbis)"
+   />
+</Target>
+
+<Target Name="_GenerateAndroidRemapNativeCode"
```

Comment:
Are we expecting these targets to run on every build? i.e no support for skipping if nothing changes between builds?

Reply:
We require the sources they generate for all builds, even if the remapping data isn't there (libxamarin-app.so ABI requirements). However, if they already generated the LLVM IR code and there are no changes, then they don't need to regenerate the sources.